### ISSUE TITLE
Treinamento RTD 

### ISSUE THREAD

**Issue by Thiago-Reis-Porto** [thacio/LLM-Notebooks#1]
Oi Thacio, eu estou tentando fazer um treinamento com RTD. Qual seria a maneira correta de converter o checkpoint do discriminador para HF? Eu me estou trabalhando com o repo original do deberta foi esse que você usou também?

**Comment by thacio** [comment]
Fala, Thiago. Usei o repo deles. Não etendi direito o seu setup, você quer treinar com o código deles ou com o hugginface? Está usando a debertina ou mdeberta como base?

No geral, os pesos são do pytorch, então a compatibilidade é quase total, mas o huggingface não atualizou a estrutura para o Deberta v3, então tem duas diferenças: os pesoas do layer de embedding (para o v3, o RTD usa uma layer 'weight', que é os pesos do gerador, mas um '_weight', que seria um específico do discriminador); e a cabeça RTD, que não foi implementada na biblioteca transformers. 
(seriam os layers 'deberta.embeddings.word_embeddings._weight' e 'deberta.embeddings.word_embeddings.weight', é só usar o torch.load para carregar, vou usar só os nomes 'weight' e '_weight' para facilitar)

Então ficaria mais ou menos assim.
Se quiser rodar com o repositório da microsoft do deberta:
- Obs: o código deles descarta o 'weight' logo de cara e faz um hook para usar o weight = embedding layer do gerador. 
- Se for com o mdeberta, funciona direto no repositório deles, eles não fizeram a junção dos layers weight + _weight (então se carregar pelo HF, a biblioteca descarta o _weight e perde infromação). No repositório do deberta, o layer 'weight' é descartado e ele usa o 'weight'do gerador, e depois inicializa o _weight com o modelo do checkpoint.
- Se for com a debertina, eu fiz a junção desses layers para não perder informação ao carregar pelo HF. Então eu fiz 'weight' = 'weight' + '_weight' no discriminador. Assim, para iniciar direto como checkpoint no repositório do deberta, o código depois descartaria o 'weight' e usaria o do gerador, e então inicializaria o _weight como zeros. É possível recuperar fácil os layers originais (o '_weight' do discriminador), seria assim:
     Carrega com o torch.load os layers do gerador e do discriminador, e cria um layer para o discriminador '_weight' = 'weight' do discriminador - 'weight' do gerador.
     Se quiser, me fala qual modelo você quer que eu devo ter em algum canto o peso original. Seria exatamente igual a esse passo acima que eu falei. Estava olhando aqui, e coloquei o gerador no [debertina-base-32k-vocab] apenas. O de 128k é de melhor performance, mas não tinha colocado o gerador lá.

Se quiser treinar na biblioteca do HF, o buraco é mais em baixo, porque eles não implementaram o v3, então precisa de implementar implementar o hook dos weights, fazer a cabeça RTD, modificar o trainer do HF para fazer o processo inteiro (gerador e discriminador ao mesmo tempo) ou utilizar dois trainers, um para cada modelo. É possível, mas esbarra em certos problemas como o de paralelização, então é melhor ir pelo repositório deles.

**Comment by Thiago-Reis-Porto** [comment]
Oi, obrigado por responder!

Estou usando o repositório original do DeBERTa e testando a versão xsmall. Fiz a conversão para o Hugging Face com o state_dict, utilizando o .weight dos word embeddings para o gerador, e testei em um STS (funcionou bem). A única modificação foi descartar os position embeddings, pois a versão DeBERTaV2 do HF não tem.

Tentei fazer o mesmo para o discriminador, mas o fine-tuning não está convergindo (o coeficiente de Pearson fica em 0 ou até negativo). Notei que o .weight do checkpoint do discriminador corresponde exatamente à soma do ._weight dele com o .weight do gerador. Agora também percebi que a loss do discriminador não está melhorando no meu pré-treinamento.

Como o dataset é grande, o dividi em vários arquivos e estou treinando por arquivo, carregando os checkpoints com a inicialização em cada etapa. Talvez o problema é que os pesos do discriminador estão sendo reiniciados a cada novo arquivo, o que está impedindo a melhora da loss.

Estou rodando o treinamento + ou - da seguinte forma por arquivo:

`python -m DeBERTa.apps.run \
  --warmup 0 \
  --dump 1000 \
  --do_train \
  --max_seq_len 512 \
  --task_name RTD \
  --data_dir /train_data/tokenized/split_1/subsplit_1 \
  --vocab_path /app/tokenizer/deberta_pt_br/deberta_pt_br_bpe.model \
  --vocab_type spm \
  --output_dir /app/train/checkpoints/xxsmall/split_1_e2/sub_1 \
  --num_train_epochs 1 \
  --model_config /app/DeBERTa/experiments/language_model/rtd_xsmall.json \
  --learning_rate 1e-4 \
  --train_batch_size 1500 \
  --eval_batch_size 1500 \
  --decoupled_training True \
  --fp16 True \
  --init_generator /app/train/checkpoints/xxsmall/split_4/sub_4/generator/pytorch.model-003659.bin \
  --init_discriminator /app/train/checkpoints/xxsmall/split_4/sub_4/discriminator/pytorch.model-003659.bin`

**Comment by Thiago-Reis-Porto** [comment]
Também quando termina um arquivo seria um como uma "epoca" dele, o código por default salvava apenas o gerador no final de uma epoca, eu coloquei um if para na última step para salvar o disc nesse casso. Quando carrega de um checkpoint também da um waring dos .weights.

`12/03/2024 10:43:52|INFO|RTD|00| Load generator from /app/train/checkpoints/xxsmall/split_1_e2/sub_3/generator/pytorch.model-007786.bin
12/03/2024 10:43:52|INFO|RTD|00| Load discriminator from /app/train/checkpoints/xxsmall/split_1_e2/sub_3/discriminator/pytorch.model-007786.bin
12/03/2024 10:43:52|WARNING|RTD|00| Load discriminator with unexptected keys: ['deberta.embeddings.word_embeddings.weight', 'deberta.embeddings.position_embeddings.weight']`

**Comment by thacio** [comment]
Pera, ainda não entendi direito, você quer fazer só o fine tunning com o discriminador ou continuar treinando ele com RTD? Se for só fine-tune, usa o hugging face mesmo, não testei com o repositório. Se for continuar o treino com RTD, aí seria no repositório.
Ou você está treinando em RTD no repositório e depois, no fine-tune, está tendo esses erros?

O xsmall é rápido e bom pra testar, mas é bem limitado, tem tarefas que ele não consegue aprender, mas não lembro exatamente quais. Eu testaria com o base para verifcar se não é problema da tarefa. O gerador costuma ser mais estável mesmo.

Esse warning é normal, é o que eu falei, o repositório deles descarta os embeddings 'weight' do discriminador e faz um hook com para usar o do gerador, só o '_weight' que é inicializado com os pesos.

Esse reinício realmente atrapalha bastante o treinamento por conta do otimizador perder o momento. É melhor ter um warmup para não divergir a rede no começo. Mas não acho que esse vai ser o problema da divergência, eu faria esses testes abaixo para tentar identificar melhor se é o modelo ou código.

Apontar o erro é bem complicado, o problema de divergência é o que mais assombra o treino de modelos, passei bastante por isso também. 
Se você estiver nesse esquema, de treino RTD depois fine-tune, e estiver divergindo neste, eu faria alguns testes.
- Fine-tune sem treino com os discriminadores do debertina xsmall (https://huggingface.co/tgsc/debertina-xsmall/tree/main), deberta base e mdeberta. Se não funcionar com esse xsmall, pode ser o problema da tarefa. Se não funcionar com os outros, pode ser problema no código.
- Depois, se funcionarem, faz um treino rápido RTD com esses modelos e testa o finetunning. Isso vai ajudar a apontar se o problema são todos os treinos, ou específico para o tamanho do modelo.
- Já que está usando o xsmall, testa o treino com fp32 para verificar a convergência.

Mas já adiantando, eu passei bastante por problema de convergência no treino e divergência no fine-tune com a Debertina.
- Inicializar com o mdeberta era treino estável, já se utilizasse o Deberta em inglês, divergia a partir de um momento do treino. O large, por exemplo, divergia sempre. Tive que começar do zero e não completei 100% do dataset, é um dos motivos da performance mais fraca dele.
- Acontcia que mesmo convergindo no treino, o modelo não aprendia no fine-tune.
- O xsmall eu treinei do zero, sem inicialização a partir de outro modelo.
- Tem discussões sobre o código do huggingface divergir em algum ponto da inferência. Na época, tinha discussão no github deles com pessoas tentando encontrar. Eu também tentei e não consegui identificar o problema.

**Comment by Thiago-Reis-Porto** [comment]
Obrigado Thacio.
Sim, a ideia seria RTD e depois fine tunning. Testei **xsmall debertina** e rolou o fine tunning. 

Usei o ASSIN2 em Similaridade Semântica. Ja o meu discriminador parece que não aprendeu, os scores de pearson e spearmen abaixo 0.1, enquanto que o gerador funciona. Por isso achei estranho. 

Não descobri ainda o que pode ser, vou tentar adicionar os pesos do "_weight" e "weight", mas se não me engano já tinha tentando. 

No pretreino toda vez que inicio um arquivo a loss está em torno de 2.6 e do gerador em torno de 2.1

**Comment by thacio** [comment]
A soma dos pesos _weight e weight não impediriam fine tunning na minha experiência.

Se vc tiver conseguido treinar com RTD e depois o finetunning com o xsmall debertina, eu suspeitaria que o problema pode ser que pesos do xsmall original estão próximos de divergir. Dá uma olhada no debertinha, nos hiperparâmetros que usaram, eles inicializaram com o mesmo modelo.

Outro teste para verificar essa hipótese pode ser começar o treino do xsmall do zero (sem iniacialização baseada em outro modelo). Se começar a ter o desempenho melhor comparado ao que foi inicializado com o xsmall, acho que seria tranquilo concluir que está divergindo por conta da inicialização dos pesos.

**Comment by Thiago-Reis-Porto** [comment]
É tinha algo quebrado em um dos checkpoints por alguma razão, peguei de um checkpoint anterior e tava convergiu o fine tunning.
Esse checkpoint estragou todos os outros posteriores kk.
Treinei mais um pouco e ta ok fine tunning.
Obrigado pela ajuda.
